cmd/registry-experimental/cmd/generate/openapi.go

```diff
@@ -0,0 +1,226 @@
+	}
+	_, err = task.client.UpdateApiSpec(ctx, &rpc.UpdateApiSpecRequest{
+		ApiSpec:      newSpec,
+		AllowMissing: true,
+	})
+	if err != nil {
+		return fmt.Errorf("failed to generate OpenAPI: %s", err)
+	}
+	return nil
+}
+
+// openAPIFromZippedProtos runs the OpenAPI generator and returns the results.
+// This uses protoc and https://github.com/google/gnostic/tree/master/cmd/protoc-gen-openapi
+// which can be installed using
+//   go install github.com/google/gnostic/cmd/protoc-gen-openapi@latest
+func openAPIFromZippedProtos(ctx context.Context, name string, b []byte) (string, error) {
+	// create a tmp directory
+	root, err := ioutil.TempDir("", "registry-protos-")
+	if err != nil {
+		return "", err
+	}
+	// whenever we finish, delete the tmp directory
+	defer os.RemoveAll(root)
+	// unzip the protos to the temp directory
+	_, err = core.UnzipArchiveToPath(b, root+"/protos")
+	if err != nil {
+		return "", err
+	}
+	return generateOpenAPIForDirectory(ctx, name, root)
+}
+
+func generateOpenAPIForDirectory(ctx context.Context, name string, root string) (string, error) {
+	// run protoc on all of the protos in the main directory
+	protos := []string{}
+	err := filepath.Walk(root+"/protos",
+		func(path string, info os.FileInfo, err error) error {
+			if err != nil {
+				return err
+			}
+			if strings.HasSuffix(path, ".proto") {
+				protos = append(protos, strings.TrimPrefix(path, root+"/"))
+			}
+			return nil
+		})
+	if err != nil {
+		return "", err
+	}
+	parts := []string{}
+	parts = append(parts, protos...)
+	parts = append(parts, "-I")
+	parts = append(parts, "protos")
+	parts = append(parts, "--openapi_out=.")
+	cmd := exec.Command("protoc", parts...)
+	cmd.Dir = root
+	log.FromContext(ctx).Debugf("Running %+v\n", cmd)
+	data, err := cmd.CombinedOutput()
+	if err != nil {
+		log.FromContext(ctx).Errorf("error %+v\n", err)
+		return "", err
+	}
+	log.FromContext(ctx).Debugf("protoc output: %s\n", string(data))
+	// attempt to read an openapi.yaml file
+	bytes, err := ioutil.ReadFile(root + "/openapi.yaml")
+	if err != nil {
+		return "", err
+	}
+	return string(bytes), err
+}
+
+// openAPIFromDiscovery runs the OpenAPI generator and returns the results.
+// This uses https://github.com/LucyBot-Inc/api-spec-converter
+// which can be installed using
+//   npm install -g api-spec-converter
+func openAPIFromDiscovery(name string, b []byte) (string, error) {
+	// create a tmp directory
+	root, err := ioutil.TempDir("", "registry-disco-")
+	if err != nil {
+		return "", err
+	}
+	// whenever we finish, delete the tmp directory
+	defer os.RemoveAll(root)
+	// write the spec to a file and run the converter
+	err = ioutil.WriteFile(root+"/discovery.json", b, 0666)
+	if err != nil {
+		return "", err
+	}
+	args := []string{"--from", "google", "--to", "openapi_3", "discovery.json"}
+	cmd := exec.Command("api-spec-converter", args...)
+	cmd.Dir = root
+	data, err := cmd.CombinedOutput()
+	if err != nil {
+		fmt.Printf("error %+v\n", err)
```

Comment:
As above, should we include a quick message about installing this?

Reply:
done (noted above)